src/nvim/api/ui.c

```diff
@@ -615,64 +642,42 @@ static bool prepare_call(UI *ui, const char *name)
     mpack_str(buf, name, strlen(name));
     data->nevents++;
     data->ncalls = 1;
-    return true;
+    return;
   }
+}
 
-  return false;
+static void send_oversized_message(UIData *data)
+{
+  if (data->temp_buf) {
+    size_t size = (size_t)(data->buf_wptr - data->temp_buf);
+    WBuffer *buf = wstream_new_buffer(data->temp_buf, size, 1, xfree);
+    rpc_write_raw(data->channel_id, buf);
+    data->temp_buf = NULL;
+    data->buf_wptr = data->buf;
+    data->nevents_pos = NULL;
+  }
 }
 
 /// Pushes data into UI.UIData, to be consumed later by remote_ui_flush().
 static void push_call(UI *ui, const char *name, Array args)
 {
   UIData *data = ui->data;
-  bool pending = data->nevents_pos;
-  char *buf_pos_save = data->buf_wptr;
-
-  bool new_event = prepare_call(ui, name);
 
   msgpack_packer pac;
   data->pack_totlen = 0;
-  data->buf_overflow = false;
+  // First determine the needed size
+  msgpack_packer_init(&pac, data, size_cb);
+  msgpack_rpc_from_array(args, &pac);
+  // Then send the actual message
+  prepare_call(ui, name, data->pack_totlen);
```

Comment:
Hmm, this works as a quick fix, but it shouldn't be needed to pack things _twice_ to figure out the buffer sizing. we want to make it safe to just simply flush the buffer at any time and then mark that we are in a state where we cannot add more events (even arglists to the same event name) to the secondary buffer(s) needed to complete the large event.

Reply:
Well, the old version did actually pack twice with memcpy  if things did not fit too. This version just determines the size on the first pack and doesn't do any actual packing. So, the overhead is just shifted to be slightly more expensive in the normal case, and less expensive in the worst case. And while I haven't done any performance profiling, the cache misses probably dominate here, and with the cache filled at the second pass it will be much faster than the first pass. I personally care most about the worst-case performance, since that could show up as frame drops in Neovide, while the normal case is always fast enough by a huge margin to not matter at all for any practical purposes.

If anything is going to be changed, I think it should be to remove the allocation of extra buffers and flush on the fly. There's nothing in the RPC protocol that requires us to allocate buffers that big. We only need to know the number of array elements, the length of the strings and so on before writing them and never the complete byte size. The stream can be flushed any time, even in the middle of a string, therefore eliminating the need for additional buffering.

But that requires bigger changes for very little benefit and increased code complexity, unless the profiling data says it's a problem. Your suggestion also probably increases at least the code complexity and development time, which at least I don't have any extra to spare.

Finally, all those micro-optimizations pales in comparison to the actual problem of sending `N(N+1) / 2` `hl_attr_define` events, where N is the number of deleted lines. That can be fixed to only require `N`, but even that's a waste for what I need. And that's why I rather spend my time working on an alternative for our use cases that doesn't need that. https://github.com/neovim/neovim/issues/23610#issuecomment-1929366416

Reply:
> If anything is going to be changed, I think it should be to remove the allocation of extra buffers and flush on the fly. There's nothing in the RPC protocol that requires us to allocate buffers that big. We only need to know the number of array elements, the length of the strings and so on before writing them and never the complete byte size. The stream can be flushed any time, even in the middle of a string, therefore eliminating the need for additional buffering.

Sure it is more something this I had in mind. I'd be happy to look at this myself, as I don't think this requires much extra complexity at all: we just need to represent the state where the "ncalls" value already has been flushed away, and thus it cannot be changed (and thus we need to start on a new `"redraw"` event for the next event, even with the same event name). Optimization should just be generally tuned to use medium sized buffers (and batching is then invoked when we can pack really small events into the _lagom_ sized buffer)

That being said, I'm fine with merging this as is as a stop-gap as-is, as it might take some time until I get into look at that and not crashing is of course priority one (especially with good testing, as you add).